statsmodels/stats/_diagnostic_other.py

```diff
@@ -0,0 +1,1043 @@
+
+
+    from statsmodels.stats.multivariate_tools import partial_project
+
+    w_sqrt = np.sqrt(weights)
+    if np.size(weights) > 1:
+        w_sqrt = w_sqrt[:,None]
+    pp = partial_project(instruments * w_sqrt, resid_deriv * w_sqrt)
+    mom_resid = pp.resid
+
+    moms_test = mom_resid * resid[:, None] * w_sqrt
+
+    # we get this here in case we extend resid to be more than 1-D
+    k_constraint = moms_test.shape[1]
+
+    # use OPG variance as in Wooldridge 1990. This might generalize
+    cov = moms_test.T.dot(moms_test)
+    diff = moms_test.sum(0)
+
+    # see Wooldridge last page in appendix
+    stat = diff.dot(np.linalg.solve(cov, diff))
+
+    # for checking, this corresponds to nobs * rsquared of auxiliary regression
+    stat2 = OLS(np.ones(nobs), moms_test).fit().ess
+
+
+    pval = stats.chi2.sf(stat, k_constraint)
+
+    return stat, pval, stat2
+
+
+def lm_robust(score, R, Ainv, B, V=None):
+    '''general formula for score/LM test
+
+    generalized score or lagrange multiplier test for implicit constraints
+
+    `r(params) = 0`, with gradient `R = d r / d params`
+
+    linear constraints are given by `R params - q = 0`
+
+    It is assumed that all arrays are evaluated at the constrained estimates.
+
+
+    Parameters
+    ----------
+    score : ndarray, 1-D
+        derivative of objective function at estimated parameters
+        of constrained model
+    constraint_matrix R : ndarray
+        Linear restriction matrix or Jacobian of nonlinear constraints
+    hessian_inv, Ainv : ndarray, symmetric, square
+        inverse of second derivative of objective function
+        TODO: could be OPG or any other estimator if information matrix
+        equality holds
+    cov_score B :  ndarray, symmetric, square
+        covariance matrix of the score. This is the inner part of a sandwich
+        estimator.
+    cov_params V :  ndarray, symmetric, square
+        covariance of full parameter vector evaluated at constrained parameter
+        estimate. This can be specified instead of cov_score B.
+
+    Returns
+    -------
+    lm_stat : float
+        score/lagrange multiplier statistic
+
+    Notes
+    -----
+
+    '''
+
+    tmp = R.dot(Ainv)
+    wscore = tmp.dot(score)  # C Ainv score
+
+    if B is None and V is None:
+        # only Ainv is given, so we assume information matrix identity holds
+        # computational short cut, should be same if Ainv == inv(B)
+        lm_stat = score.dot(Ainv.dot(score))
+    else:
+        # information matrix identity does not hold
+        if V is None:
+            inner = tmp.dot(B).dot(tmp.T)
+        else:
+            inner = R.dot(V).dot(R.T)
+
+        #lm_stat2 = wscore.dot(np.linalg.pinv(inner).dot(wscore))
+        # Let's assume inner is invertible, TODO: check if usecase for pinv exists
+        lm_stat = wscore.dot(np.linalg.solve(inner, wscore))
+
+    return lm_stat#, lm_stat2
+
+
+def lm_robust_subset(score, k_constraints, score_deriv, cov_score):
```

Comment:
I guess it would be better of k_constraints allows an index as option, instead of assuming sorted by Null hypothesis